src/librustc_trans/trans/mir/rvalue.rs

```diff
@@ -45,8 +45,36 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
                 bcx
             }
 
-            mir::Rvalue::Cast(..) => {
-                unimplemented!()
+            mir::Rvalue::Cast(mir::CastKind::Unsize, ref operand, cast_ty) => {
+                if common::type_is_fat_ptr(bcx.tcx(), cast_ty) {
+                    let (bcx, temp) = self.trans_rvalue_operand(bcx, rvalue);
+                    self.store_operand(bcx, lldest, temp);
+                    return bcx;
+                }
+
+                // Unsize of a nontrivial struct. I would prefer for
+                // this to be eliminated by MIR translation, but
+                // `CoerceUnsized` can be passed by a where-clause,
+                // so the (generic) MIR may not be able to expand it.
+                let operand = self.trans_operand(bcx, operand);
+                match operand.val {
+                    OperandValue::FatPtr(..) => unreachable!(),
+                    OperandValue::Imm(llval) => {
+                        // ugly alloca.
+                        debug!("trans_rvalue: creating ugly alloca");
+                        let lltemp = base::alloc_ty(bcx, operand.ty, "__unsize_temp");
```

Comment:
Question out of curiosity: Is there any reason this should be alloca and cannot be regular LLVM temporary variable?